### PR TITLE
[SPARK-18947][SQL] SQLContext.tableNames should not call Catalog.listTables

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

It's a huge waste to call `Catalog.listTables` in `SQLContext.tableNames`, which only need the table names, while `Catalog.listTables` will get the table metadata for each table name.

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @gatorsmile

**Comment by gatorsmile** [comment]
The same issue also exists in [getTableNames](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala#L278). Could we also fix there?

**Comment by gatorsmile** [comment]
LGTM except the above comment

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 1]
```diff
@@ -747,7 +747,7 @@ class SQLContext private[sql](val sparkSession: SparkSession)
    * @since 1.3.0
    */
   def tableNames(): Array[String] = {
-    sparkSession.catalog.listTables().collect().map(_.name)
+    sessionState.catalog.listTables(sessionState.catalog.getCurrentDatabase).map(_.table).toArray
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: can you explain how it solves the issue?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: The original one is using [`sparkSession.catalog.listTables`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala#L83-L99); the new one is using [`sessionState.catalog.listTables`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala#L600-L632). The implementation of these two function calls are different. The later one is using [getTablesByPattern](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java#L1371-L1381).

To be honest, I did not notice the difference at the beginning. : )

File: sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala [hunk 2]
```diff
@@ -276,11 +276,12 @@ private[sql] object SQLUtils extends Logging {
   }
 
   def getTableNames(sparkSession: SparkSession, databaseName: String): Array[String] = {
-    databaseName match {
-      case n: String if n != null && n.trim.nonEmpty =>
-        sparkSession.catalog.listTables(n).collect().map(_.name)
+    val db = databaseName match {
+      case _ if databaseName != null && databaseName.trim.nonEmpty =>
+        databaseName.trim
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: uh... not sure whether we should support triming. So far, when we do something like
```Scala
session.tableNames("default ")
```

It reports the error:
```
Database 'default ' not found;
```

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: ok let me keep the previous behavior, although it's weird(check `...trim.nonEmpty` but not use the trimmed database name)....

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala]: : ) Yeah

**Comment by gatorsmile** [comment]
LGTM again pending test

**Comment by gatorsmile** [comment]
retest this please

**Pull request closed without merging**